File: Game.Application/GameJS/src/utils/LibNoiseUtils.ts
import { libnoise } from "libnoise";
import { construct, DataDrivenInput, DataDrivenConstructorRecords } from "./DataDrivenComposition";
import { NonCohesiveNoiseGenerator } from "../terrain-generation/NonCohesiveNoiseGenerator";

export const DEFAULT_PERLIN_LACUNARITY = 2;
export const DEFAULT_PERLIN_SEED = 0;
export const DEFAULT_PERLIN_OCTAVE_COUNT = 8;
export const DEFAULT_PERLIN_FREQUENCY = 1;
export const DEFAULT_PERLIN_PERSISTENCE = 0.5;

class Perlin extends libnoise.generator.Perlin {
  constructor(frequency: number, lacunarity: number, persistence: number, octaves: number, seed: number) {
    super(frequency, lacunarity, persistence, octaves, seed, libnoise.QualityMode.MEDIUM);
  }
}

class RidgedMultifractal extends libnoise.generator.RidgedMultifractal {
  constructor(frequency: number, lacunarity: number, octaves: number, seed: number) {
    super(frequency, lacunarity, octaves, seed, libnoise.QualityMode.MEDIUM);
  }
}

const dataDrivenConstructors: DataDrivenConstructorRecords<any, libnoise.ModuleBase> = {
  "generator.Perlin": Perlin,
  "generator.RidgedMultifractal": RidgedMultifractal,
  "generator.NonCohesiveNoiseGenerator": NonCohesiveNoiseGenerator,
  "operator.Abs": libnoise.operator.Abs,
  "operator.Add": libnoise.operator.Add,
  "operator.Clamp": libnoise.operator.Clamp,
  "generator.Const": libnoise.generator.Const,
  "operator.Multiply": libnoise.operator.Multiply,
  "operator.Translate": libnoise.operator.Translate,
  "operator.Scale": libnoise.operator.Scale,
};

export function dataDrivenNoise<T extends DataDrivenInput<number, libnoise.ModuleBase>>(target: T) {
  return construct(target, dataDrivenConstructors);
}

File: Game.Domain/LibNoiseUtils.cs
using LibNoise;
using LibNoise.Generator;
using LibNoise.Operator;
using System;
using System.Collections.Generic;
using System.Text;

namespace Game.Domain
{
    public static class LibNoiseUtils
    {
        public const float DEFAULT_PERLIN_LACUNARITY = 2f;
        public const int DEFAULT_PERLIN_SEED = 0;
        public const int DEFAULT_PERLIN_OCTAVE_COUNT = 6;
        public const float DEFAULT_PERLIN_FREQUENCY = 1f;
        public const float DEFAULT_PERLIN_PERSISTENCE = 0.5f;

        public const float Y_OFFSET = 1 / 400f;

        public static ModuleBase InitializePerlin(float lacunarity = DEFAULT_PERLIN_LACUNARITY, int seed = 0, int octaveCount = DEFAULT_PERLIN_OCTAVE_COUNT, float frequency = DEFAULT_PERLIN_FREQUENCY)
        {
            return new Translate(0, Y_OFFSET, 0, 
                new Clamp(0, 1, 
                    new Add(
                        new Multiply(
                            new Perlin(frequency, lacunarity, DEFAULT_PERLIN_PERSISTENCE, octaveCount, seed, QualityMode.Medium), 
                            new Const(1 / 1.77f)
                        ), 
                        new Const(0.5f)
                    )
                )
            );
        }

        class NormalRidge : ModuleBase
        {
            private readonly ModuleBase original;

            public NormalRidge(ModuleBase original)
            {
                this.original = original;
            }

            public override double GetValue(double x, double y, double z)
            {
                var val = Utils.Clamp(original.GetValue(x, y, z), 0, 1);
                return 1 - val * (1 - val) * 4;
            }
        }
    }
}
